### PR TITLE
Permit trivial solves for floating point equality in ShapeEnv

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126074
* __->__ #125915
* #125325

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -4570,10 +4570,20 @@ def _smart_symbol_sort(x):
...
-                if isinstance(lhs, sympy.Symbol) and free_unbacked_symbols(lhs) and not free_unbacked_symbols(rhs):
-                    # short-circuit when no solving is needed
+                def trivial_solve(lhs, rhs):
+                    if isinstance(lhs, sympy.Symbol):
+                        if free_unbacked_symbols(lhs) and not free_unbacked_symbols(rhs):
+                            return True
+                        if symbol_is_type(lhs, SymT.FLOAT):
+                            return True
+                        # TODO: Maybe trivial solutions for int should also be
+                        # done?
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:4580]: We are doing this in the else... just for the lhs. Oh well.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -4570,10 +4570,20 @@ def _smart_symbol_sort(x):
                 # references u2 and u3 prior to them actually being bound at
                 # runtime.  It's pretty inconvenient to setup control
                 # dependencies for substitutions, so ban it entirely.
-                if isinstance(lhs, sympy.Symbol) and free_unbacked_symbols(lhs) and not free_unbacked_symbols(rhs):
-                    # short-circuit when no solving is needed
+                def trivial_solve(lhs, rhs):
+                    if isinstance(lhs, sympy.Symbol):
+                        if free_unbacked_symbols(lhs) and not free_unbacked_symbols(rhs):
+                            return True
+                        if symbol_is_type(lhs, SymT.FLOAT):
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:4577]: do we need any sort of type check on the rhs?

**Review by lezcano — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:4577]: No, I think lhs being a float guarantees rhs is float

Comment (lezcano) [reply torch/fx/experimental/symbolic_shapes.py:4577]: Ah, yes, writing asserts in comments. Classic :P

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:4523]: I mean, it's not trivial if you ran try_solve on it.

**Comment by ezyang** [comment]
oops, I guess I can't reorder this stack

**Pull request closed without merging**